polar-core/src/resource_block.rs

```diff
@@ -280,21 +282,43 @@ impl ResourceBlocks {
         resource: Term,
         declarations: Declarations,
         shorthand_rules: Vec<ShorthandRule>,
-    ) {
-        self.declarations.insert(resource.clone(), declarations);
+    ) -> Vec<ValidationError> {
+        let mut errors = vec![];
+
+        // Merge existing declarations if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.declarations.get_mut(&resource) {
+            for (key, new) in declarations {
+                if let Some(existing) = existing.insert(key.clone(), new.clone()) {
+                    if existing != new {
+                        errors.push(ValidationError::DuplicateResourceBlockDeclaration {
+                            resource: resource.clone(),
+                            declaration: key.clone(),
+                            existing,
+                            new: new.clone(),
+                        });
+                    }
+                }
+            }
+        } else {
+            // or insert a new set of declarations for the resource type
+            self.declarations.insert(resource.clone(), declarations);
+        }
+
+        // Merge existing shorthand rules if we are reopening a resource block, otherwise add new
         self.shorthand_rules
-            .insert(resource.clone(), shorthand_rules);
+            .entry(resource.clone())
+            .and_modify(|existing| existing.extend(shorthand_rules.clone()))
+            .or_insert(shorthand_rules);
```

Comment:
Just a note that this merge might result in duplicate shorthand rules that'll get rewritten into duplicate longhand rules. That behavior's more correct than silently de-duping, but just a reminder that we'll want to circle back and error (or at least warn) on duplicate rules in a follow-up task.